edb/server/compiler_pool/server.py

```diff
@@ -0,0 +1,547 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2022-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+
+import asyncio
+import collections
+import functools
+import logging
+import os
+import pickle
+import time
+import traceback
+import typing
+
+import click
+import immutables
+
+from edb.common import debug
+from edb.common import markup
+
+from .. import args as srvargs
+from . import amsg
+from . import pool as pool_mod
+from . import worker as worker_mod
+from . import state as state_mod
+
+
+_client_id_seq = 0
+_tx_state_id_seq = 0
+logger = logging.getLogger("edb.server")
+
+
+def next_tx_state_id():
+    global _tx_state_id_seq
+    _tx_state_id_seq = (_tx_state_id_seq + 1) % (2 ** 63 - 1)
+    return _tx_state_id_seq
+
+
+class PickledState(typing.NamedTuple):
+    user_schema: typing.Optional[bytes]
+    reflection_cache: typing.Optional[bytes]
+    database_config: typing.Optional[bytes]
+
+    def diff(self, other: PickledState):
+        return PickledState(
+            None
+            if self.user_schema is other.user_schema
+            else self.user_schema,
+            None
+            if self.reflection_cache is other.reflection_cache
+            else self.reflection_cache,
+            None
+            if self.database_config is other.database_config
+            else self.database_config,
+        )
+
+
+class ClientSchema(typing.NamedTuple):
+    dbs: immutables.Map[str, PickledState]
+    global_schema: typing.Optional[bytes]
+    instance_config: typing.Optional[bytes]
+    dropped_dbs: tuple
+
+    def diff(self, other: ClientSchema):
+        dropped_dbs = tuple(
+            dbname for dbname in other.dbs if dbname not in self.dbs
+        )
+        dbs: immutables.Map[str, PickledState] = immutables.Map()
+        for dbname, state in self.dbs.items():
+            other_state = other.dbs.get(dbname)
+            if other_state is None:
+                dbs = dbs.set(dbname, state)
+            elif state is not other_state:
+                dbs = dbs.set(dbname, state.diff(other_state))
+        return ClientSchema(
+            dbs,
+            None
+            if self.global_schema is other.global_schema
+            else self.global_schema,
```

Comment:
nit: I personally find this hard to read, I'd just move this into a variable.

Reply:
Right!